lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
-          else
-            return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                    UsedRegisters.begin()->Location,
-                                    TheBinary);
-        } else {
-          return std::nullopt;
-        }
+        // Otherwise return a `void *` pointer as the return value type.
+        static constexpr auto VoidKind = model::PrimitiveTypeKind::Void;
+        auto Result = ensurePrimitiveTypeIsAvailable(VoidKind,
+                                                     0,
+                                                     Binary,
+                                                     TypeDependencies);
+        using model::Qualifier;
+        Result.Qualifiers.emplace_back(Qualifier::createPointer(Architecture));
+        return Result;
+      }
+    }
+
+    // TODO: there are some cases where preserving the total structure of
+    // the return value can be benefitial, this needs more work.
+    // For now, just determine the total size of the return value and if it
+    // fits into a primitive value - emit it, otherwise show those values as
+    // a struct (/note: positional ABIs need a whole lot of extra love in this
+    // regard).
+    llvm::SmallVector<std::optional<model::QualifiedType>, 4> ResultingTypes;
+    bool MustUseTheNextOne = false;
+    auto AllowedRange = llvm::enumerate(llvm::reverse(Allowed));
+    for (auto Pair : AllowedRange) {
+      size_t Index = Allowed.size() - Pair.index() - 1;
+      model::Register::Values Register = Pair.value();
+      auto UsedIterator = Used.find(Register);
+      if (UsedIterator != Used.end())
+        ResultingTypes.emplace_back(UsedIterator->Type);
+      else if (MustUseTheNextOne)
+        ResultingTypes.emplace_back(std::nullopt);
+
+      MustUseTheNextOne = MustUseTheNextOne || (UsedIterator != Used.end());
+    }
+
+    revng_assert(ResultingTypes.size() != 0);
+    if (ResultingTypes.size() == 1) {
+      // Short-circuit for a single register return value.
+      auto Kind = model::Register::primitiveKind(Allowed.at(0));
+      if (ResultingTypes.front() != std::nullopt) {
+        return ensureTypeIsAvailable(*ResultingTypes.front(),
+                                     Allowed.at(0),
+                                     Binary,
+                                     TypeDependencies);
+      } else {
+        // No information about the type of this register, use the default.
+        return ensureTypeIsAvailable(model::QualifiedType{},
+                                     Allowed.at(0),
+                                     Binary,
+                                     TypeDependencies);
+      }
+    } else if (ResultingTypes.size() == 2) {
+      // Short-circuit for a double register return value.
+      if (Allowed.size() < 2) {
+        // The ABI doesn't support double register return values.
+        // There's a good chance the ABI was not detected correctly.
+        return std::nullopt;
       }
```

Comment:
I don't think this can happen: `ResultingTypes.size()` is at most `Allowed.size()`.
Maybe you can `assert`?

Reply:
The case where I imagine it happening is when ABI is miss-detected or the user has set a wrong one and tries to convert a function under the ABI that allows double-return values (`RawFunctionType` actually has multiple registers listed) into a CABI with ABI that does not (with only a single register for return values, for example) - even though `detectABI` analysis should in theory prevent this, I'd rather stay safe and keep this a soft fail instead.

Reply:
What I mean is that it's technically impossible to enter this `if`.
It can never happen that `ResultingTypes.size() == 2` *and* `Allowed.size() < 2`.
`Allowed.size()` is an upper bound to the size of `ResultingTypes`, unless I'm missing something.

Reply:
This is also obsoleted thanks to using the `RegisterStateDeductions` 🎉.